Create arc-23: Append the hash of the contract description to the compiled application's bytes (#80)
algorandfoundation/ARCs: ARCs/arc-0023.md

## Abstract

The goal of this convention is to standardize the process of verifying and interacting with smart contracts.

## Specification

The key words "**MUST**", "**MUST NOT**", "**REQUIRED**", "**SHALL**", "**SHALL NOT**", "**SHOULD**", "**SHOULD NOT**", "**RECOMMENDED**", "**MAY**", and "**OPTIONAL**" in this document are to be interpreted as described in [RFC-2119](https://www.ietf.org/rfc/rfc2119.txt).

> Comments like this are non-normative.

## Overview

By appending the hash of the contract description, any user with access to the blockchain could easily verify the ABI of an application and interact with it.

> The goal is to have an off-chain ABI repository, where entries are indexed by the hash of the contract description.

### Hash

The hash is defined as follow:
```
SHA-512/256("arc0023" || <content of the metadata JSON file>)
```
In other words, it is the SHA-512/256 hash of the content of the metadata JSON file prefixed by the 7-byte string `arc0023`.
### Appending

> The encoded object bytes are appended to the compiled application as a [bytecblock](https://developer.algorand.org/docs/get-details/dapps/avm/teal/opcodes/#bytecblock-bytes) containing 1 byte constant which is the encoded object.

Reviewer: For programs that do not use `return`, I believe `pushbytes` is an issue since it will change the semantic.
For example the basic program:

```
#pragma version 5
int 1
```

changes semantic when adding the `pushbytes`

> The encoded object bytes are appended to the compiled application as a [bytecblock](https://developer.algorand.org/docs/get-details/dapps/avm/teal/opcodes/#bytecblock-bytes) containing 1 byte constant which is the encoded object.

Reviewer 2: Good point, perhaps the developer can choose between these options depending on whether they end with a `return` or not?

> The encoded object bytes are appended to the compiled application as a [bytecblock](https://developer.algorand.org/docs/get-details/dapps/avm/teal/opcodes/#bytecblock-bytes) containing 1 byte constant which is the encoded object.

Reviewer 3: I think the current assembler is going to get upset with a second bytecblock. I think that if it sees a bytecblock, it will not insert its own. I suppose that could be changed. Oh perhaps `pushbytes XXX; pop` (though that has the downside of costs 2 opcodes if you had planned on running off the end of the program with no opcodes to spare.

> The encoded object bytes are appended to the compiled application as a [bytecblock](https://developer.algorand.org/docs/get-details/dapps/avm/teal/opcodes/#bytecblock-bytes) containing 1 byte constant which is the encoded object.

Author: @jannotti you're right, the assembler is ignoring its own `bytecblock`.
I guess the better solution is using `pushbytes` following a `return`. If `return` is already the last command then there's no need for another one of-course .

> The encoded object bytes are appended to the compiled application as a [bytecblock](https://developer.algorand.org/docs/get-details/dapps/avm/teal/opcodes/#bytecblock-bytes) containing 1 byte constant which is the encoded object.

Reviewer 3: It's weird, but this could also change semantics. It is currently ok to jump to the PC after the last byte as a way to end the program. I doubt anyone does it, but it's possible.

                        ],
                        [
                            Txn.application_args[0] == Bytes('base16', '766083a7'),  # multiply(uint64,uint64)uint64
                            Seq([
                                Log(
                                    Concat(
                                        Bytes('base16', '151f7c75'),
                                        Itob(Mul(Btoi(Txn.application_args[1]), Btoi(Txn.application_args[2])))
                                    )
                                ),
                                Approve()
                            ])
                        ],

                    )
                ]
            ), mode=Mode.Application, version=5
        )
```

With the following python script we can generate the hash and append it to the compiled application, according to arc-23:
```python
import hashlib
import msgpack  # Python 3.8.3, msgpack==1.0.2

def sha512_256(data: bytes):
    return hashlib.new(name='sha512_256', data=data).digest()

def hash_contract_descriptor(contract_descriptor: bytes):
    return sha512_256(b'arc0023' + contract_descriptor)

def encode_arc23_object(contract_descriptor: bytes):

>     return msgpack.packb({'arc23': hash_contract_descriptor(contract_descriptor)})

Reviewer 2: The message you're adding to the end of the program is not just a hash, it's the msgpack encoding of the object `{"arc23": <32-byte-hash>}`? Is the assumption that other ARCs might want to add similar things, in which case more keys on that object would be present, such as `{"arc23": <32-byte-hash>, "arc999": <32-byte-hash>}`? If that's the reason, here are some downsides:
* This is much less space efficient, since every "standard" that a contract wants to support takes _at least_ 32 bytes each.
* Since this object gets encoded as a variable-length byte string, it will be harder for clients to properly decode it. You can't just look at the last `N` bytes, you need to actually decode every opcode in the program, look at the last one, and be able to decode msgpack. That's a lot of requirements for every client to follow.

Consider this slightly different approach:
* Instead of each "standard" being applied independently to a program, create a unified JSON object that contains all the information from each standard.
* This object could have keys for each standard, and the values would be the full metadata for that standard. For example:
```json
{
  "arc23": { "name": "..." },
  "arc999": { "field_specific_to_arc999": "value..." }
}
```
* To apply this information to the contract, hash the entire object and place only the 32-byte hash at the end of the program using `bytecblock` or `pushbytes`. This way, clients need only look at the last 32-bytes of a program to prove that it implements a set of standards. No msgpack decoding or program op parsing is required.

Reviewer 3: Another option, which impacts the earlier discussion about using `bytecblock` vs `pushbytes`, might be to leverage the structure of the bytecblock for some of the key/value properties you're looking for, rather than msgpack.  So, consider:
```
bytecblock "arc-23" 0x32BYTEHEX "arc-XX" 0xSOMEOTHERHASH NNN
where NNN is a 4 byte, bigendian number that says how big the entire bytecblock instruction is.

That would allow clients to find the instruction without diassembling bytecode or understanding messagepack (though they do need to unpack the uvarint encoding of bytecblock). OR if they DO use a disassembler, everything pretty much gets unpacked nicely for them without messagepack or explicit knowledge of uvarint, etc.

Reviewer 3: And yet another option is to insert this special bytecblock after the version byte at the start of the program. That would be safe because jumps are relative, so it would push everything forward.  It would be easier to parse, because it starts at a known location.

The only downside I can think of is that if you disassemble such a modified program, it will have two bytecblocks, which the assembler _might_ complain about.  But if went this way, I would change the assembler if needed.

## Abstract

The goal of this convention is to standardize the process of verifying and interacting with smart contracts.

## Specification

The key words "**MUST**", "**MUST NOT**", "**REQUIRED**", "**SHALL**", "**SHALL NOT**", "**SHOULD**", "**SHOULD NOT**", "**RECOMMENDED**", "**MAY**", and "**OPTIONAL**" in this document are to be interpreted as described in [RFC-2119](https://www.ietf.org/rfc/rfc2119.txt).

> Comments like this are non-normative.

## Overview

By appending the hash of the contract descriptor, any user with access to the blockchain could easily verify the ABI of an application and interact with it.

> The goal is to have an off-chain ABI repository, where entries are indexed by the hash of the contract descriptor.

TEAL compilers are responsible for appending an ecnoded object contating the hash and some metadata.

### Hash

Any hash function can be used, but the default and **RECOMMENDED** is `SHA-512/256`.
The input to the hash function **MUST** be the canonical JSON object of the contract descriptor.

> 

Author: do we prefer `metadata JSON file` or `contract descriptor`?

> 

Author: using `contract description`

### Appending

The encoded object bytes are appended to the compiled application as a [bytecblock](https://developer.algorand.org/docs/get-details/dapps/avm/teal/opcodes/#bytecblock-bytes) containing 1 byte constant which is the encoded object.

> 

Reviewer: ```suggestion
> The reason to use `bytecblock` is that adding a `bytecblock` opcode at the end of a TEAL program does not change the semantics of the program, as long as: opcodes are properly aligned, there is no jump after the last position (that would make the program fails without `bytecblock`), and there is enough space left to add the opcode.
```

The size of the compiled application + the bytecblock should not exceed the maximum size of a compiled application according to the latest consensus parameters supported by the compiler.

### Encoding

> The encoded object is a MessagePack encoding of a mapping object from hash functions to hash values. The default is:

Reviewer: ```suggestion
The encoded object is a MessagePack encoding of the following object:
```
